oak_runtime/src/node/http/mod.rs

```diff
@@ -78,7 +78,16 @@ impl hyper::server::accept::Accept for TlsServer<'_> {
         mut self: Pin<&mut Self>,
         cx: &mut Context,
     ) -> Poll<Option<Result<Self::Conn, Self::Error>>> {
-        Pin::new(&mut self.acceptor).poll_next(cx)
+        let connection = Pin::new(&mut self.acceptor).poll_next(cx);
+        match connection {
+            // In case of error (e.g., CertificateUnknown error), return Poll::Pending, to avoid
+            // early termination of the server.
+            Poll::Ready(Some(Err(e))) => {
+                log::error!("Error when processing TLS stream: {:?}", e);
+                Poll::Pending
```

Comment:
What is the difference between this and `Poll::Ready(None)`?

In fact, I don't really understand why `Pending` works here, given what it says here https://doc.rust-lang.org/nightly/core/task/enum.Poll.html#variant.Pending :

> When a function returns Pending, the function must also ensure that the current task is scheduled to be awoken when progress can be made.

Reply:
I also expected `Poll::Ready(None)` to work, but it does not. Which is strange, because as far as I can see, this is where `poll_accept` is called: 

https://github.com/hyperium/hyper/blob/66fc127c8d4f81aed9300c9d0f13246b8206067a/src/server/conn.rs#L773

If I return `Poll::Ready(None)`, the server terminates as before. 
I have added the call to `cx.waker().clone().wake()` before returning `Poll::Pending`, but I am not sure what impact it has.

Reply:
I don't think that's what is expected to happen (at least, waking up the task before calling Pending does not match my mental model). Could you try and see what is happening exactly, and if this is the way to go, add more comments to explain how you arrived to your conclusion?

Reply:
Having looked at the Hyper code, I think returning `Poll::Pending` is appropriate. It treats `poll_accept` similarly to an async stream. Returning `Poll::Ready(None)` (or an error) signals the end of the stream, so it returns ouf of the loop.

I don't think `cx.waker().clone().wake()` does anything useful. `Poll::Pending` works because the context is passed to the underlying acceptor, which would call `wake` on the `waker` when there is a new incoming connection. From the server's point of view, it was woken up when there was a new connection, but the connection failed so there is nothing for it to do but wait for the next connection.

Reply:
Thanks for the nice explanation @conradgrobler. This is my understanding too (based on what I observed), but I don't see where in the code `Poll::Ready(None)` signals the end of the stream. Do you know where in the code it happens?

Reply:
I believe it happens at https://github.com/hyperium/hyper/blob/66fc127c8d4f81aed9300c9d0f13246b8206067a/src/server/conn.rs#L852

Reply:
But I expect to reach that line with both `Poll::Pending` and `Poll::Ready(None)`. Does not the code reach that line with `Poll::Pending` too?

Reply:
The `ready!` macro in https://github.com/hyperium/hyper/blob/66fc127c8d4f81aed9300c9d0f13246b8206067a/src/server/conn.rs#L843 means that the function would return earlly with `Poll::Pending` when it receives `Poll::Pending`. This in turn will cause the server to return `Poll::Pending` from its implementation of `poll`, which would in effect cause a wait until there is a new connection.

Reply:
Oh! Of course! `ready!` is a macro. Thanks.

Reply:
Also see https://docs.rs/hyper/0.13.7/src/hyper/server/tcp.rs.html#151

BTW can we just use `AddrIncoming` from that crate?

Reply:
Is `AddrIncoming` handling TLS connections?

Reply:
uhmm no, you are right, my bad :)

Reply:
Why would `ready!` not work here too then, given this method seems quite similar in structure to `AddrIncoming` anyways?

Reply:
The logic here is different from `ready!`. `ready!` would return an error, we want to return `Poll::Pending`. Also, I don't think `ready!` is public (its definition does not have a `#[macro_export]` attribute).

Reply:
I thought this is it, I'm not sure why it wouldn't work

https://docs.rs/futures-util/0.3.5/futures_util/macro.ready.html

Reply:
I think the `ready!` macro used in hyper is this one: 
https://github.com/hyperium/hyper/blob/master/src/common/mod.rs

Reply:
I meant that, in conjunction with some additional error handling logic, to be clear. Anyways, if it doesn't work, feel free to ignore.

Reply:
> I think the `ready!` macro used in hyper is this one:
> https://github.com/hyperium/hyper/blob/master/src/common/mod.rs

I understand that, but why can we not use https://docs.rs/futures-core/0.3.5/src/futures_core/task/poll.rs.html#5-11 (which AFAICT has an equivalent implementation) and make the code look more similar to https://docs.rs/hyper/0.13.7/src/hyper/server/tcp.rs.html#98 for instance? To me it seems the pattern is quite similar. Anyways, we can try that out separately perhaps.

Reply:
Yes. The pattern is similar. I can try that.

Reply:
As discussed, I'll leave it as is (without using `ready!`) for now, since the `poll_accept` does not need to process the returned value apart from replacing it with `Poll::Pending` in case of an error. If in the future, we need a more complicated logic, we could use `ready!`.